File: cli-ios/RemoteCoding/Rendering/Renderers/ContentRenderers.swift
import SwiftUI

/// PROTOCOL §5.2 —— 内容 / 叶子组件(深色皮肤)。

struct TextRenderer: View {
    let component: Component
    var body: some View {
        let p = component.props
        let text = p.string("text")
        let isMarkdown = p.bool("markdown")
        if isMarkdown {
            MarkdownText(text: text)   // 完整 markdown 渲染(深色主题)
        } else {
            plain(p, text)
        }
    }

    @ViewBuilder
    private func plain(_ p: JSONValue, _ text: String) -> some View {
        let style = p.string("style", default: "body")
        let mono = p.bool("mono")
        let bold = p.bool("bold")
        let size = Self.size(style)
        let weight: Font.Weight = (bold || style == "heading") ? .semibold : .regular
        let font: Font = mono
            ? .system(size: size, weight: weight, design: .monospaced)
            : .system(size: size, weight: weight)
        let fg: Color = p["color"]?.stringValue.map { Theme.named($0, default: Theme.text) }
            ?? (style == "caption" ? Theme.textSec : Theme.text)
        // fill=true 让文本块撑满一行(默认);行内小元素(如要点圆点)设 fill=false 保持紧凑。
        let fill = p.bool("fill", default: true)

        Text(text)
            .font(font)
            .foregroundStyle(fg)
            .multilineTextAlignment(.leading)
            .frame(maxWidth: fill ? .infinity : nil, alignment: .leading)
    }

    static func size(_ style: String) -> CGFloat {
        switch style {
        case "heading": return 17
        case "caption": return 13
        default:        return 15
        }
    }
}

/// 会话文本块。agent → 左侧悬浮卡(配合外侧头像);user → 右对齐蓝气泡。
struct BubbleRenderer: View {
    let component: Component
    var body: some View {
        let p = component.props
        let textStr = p.string("text")
        let isMarkdown = p.bool("markdown")
        let isUser = p.string("role", default: "agent") == "user"

        let content = Group {
            if isMarkdown, let attr = try? AttributedString(markdown: textStr) {
                Text(attr)
            } else {
                Text(textStr)
            }
        }
        .font(.system(size: 15))
        .foregroundStyle(Theme.text)
        .multilineTextAlignment(.leading)
        .lineSpacing(2)

        if isUser {
            // 右对齐由 TaskDetailView 的 user 分支统一处理,这里只出气泡本体。
            content
                .padding(.horizontal, 14).padding(.vertical, 10)
                .background(Theme.blueBtn.opacity(0.22))
                .clipShape(RoundedRectangle(cornerRadius: 16))
        } else {
            content
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(14)
                .cardStyle()
        }
    }
}

struct CodeRenderer: View {
    let component: Component
    @State private var copied = false

    var body: some View {
        let code = component.props.string("code")
        let language = component.props["language"]?.stringValue
        let copyable = component.props.bool("copyable", default: true)

        VStack(alignment: .leading, spacing: 6) {
            if language != nil || copyable {
                HStack {
                    if let language { Text(language).font(.system(size: 11)).foregroundStyle(Theme.textTer) }
                    Spacer()
                    if copyable {
                        Button {
                            UIPasteboard.general.string = code
                            copied = true
                        } label: {
                            Label(copied ? "已复制" : "复制", systemImage: copied ? "checkmark" : "doc.on.doc")
                                .font(.system(size: 11)).foregroundStyle(Theme.textSec)
                        }.buttonStyle(.plain)
                    }
                }
            }
            // 自动换行显示完整命令(不再横向滚动只露开头)—— 审批时能一眼看清整条命令。
            Text(code).font(.system(.footnote, design: .monospaced)).foregroundStyle(Theme.text)
                .frame(maxWidth: .infinity, alignment: .leading)
                .textSelection(.enabled)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.field)
        .overlay(RoundedRectangle(cornerRadius: 10).stroke(Theme.stroke, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 10))
    }
}

struct BadgeRenderer: View {
    let component: Component
    var body: some View {
        let text = component.props.string("text")
        let color = Theme.named(component.props["color"]?.stringValue, default: Theme.textSec)
        Text(text)
            .font(.system(size: 12, weight: .medium))
            .padding(.horizontal, 9).padding(.vertical, 4)
            .background(color.opacity(0.16))
            .foregroundStyle(color)
            .clipShape(Capsule())
    }
}

struct KeyValueRenderer: View {
    let component: Component
    var body: some View {
        let items = component.props["items"]?.arrayValue ?? []
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(items.enumerated()), id: \.offset) { _, item in
                HStack(alignment: .top) {
                    Text(item.string("k")).font(.system(size: 13)).foregroundStyle(Theme.textSec)
                    Spacer()
                    Text(item.string("v")).font(.system(size: 13)).foregroundStyle(Theme.text)
                        .multilineTextAlignment(.trailing)
                }
            }
        }
    }
}

struct ProgressRenderer: View {
    let component: Component
    var body: some View {
        let label = component.props["label"]?.stringValue
        let value = component.props.double("value")
        VStack(alignment: .leading, spacing: 6) {
            if let label { Text(label).font(.system(size: 13)).foregroundStyle(Theme.textSec) }
            if let value {
                ProgressView(value: max(0, min(1, value))).tint(Theme.blue)
            } else {
                ProgressView().tint(Theme.blue)   // 无限转圈
            }
        }
    }
}

struct ImageRenderer: View {
    let component: Component
    var body: some View {
        let p = component.props
        // 优先内联 base64(Mac 本地图片经 agent 缩略图内联);否则走 url。
        if let dataStr = p["data"]?.stringValue,
           let data = Data(base64Encoded: dataStr), let ui = UIImage(data: data) {
            Image(uiImage: ui).resizable().scaledToFit()
                .frame(maxWidth: 240, maxHeight: 300)
                .clipShape(RoundedRectangle(cornerRadius: 16))
                .overlay(RoundedRectangle(cornerRadius: 16).stroke(Theme.stroke.opacity(0.5), lineWidth: 0.5))
                .shadow(color: .black.opacity(0.25), radius: 6, y: 2)
        } else if let urlString = p["url"]?.stringValue ?? p["thumbUrl"]?.stringValue,
                  let url = URL(string: urlString) {
            AsyncImage(url: url) { phase in
                switch phase {
                case .success(let img): img.resizable().scaledToFit()
                case .failure: Image(systemName: "photo").foregroundStyle(Theme.textTer)
                default: ProgressView().tint(Theme.blue)
                }
            }
            .frame(maxWidth: 220, maxHeight: 280)
            .clipShape(RoundedRectangle(cornerRadius: 14))
        } else {
            Image(systemName: "photo").font(.system(size: 28)).foregroundStyle(Theme.textTer)
                .frame(width: 120, height: 90).background(Theme.cardHi)
                .clipShape(RoundedRectangle(cornerRadius: 12))
        }
    }
}

struct DiffRenderer: View {
    let component: Component
    var body: some View {
        let filename = component.props["filename"]?.stringValue
        let hunks = component.props["hunks"]?.arrayValue ?? []
        VStack(alignment: .leading, spacing: 0) {
            if let filename {
                Text(filename).font(.system(size: 12, design: .monospaced)).foregroundStyle(Theme.textSec)
                    .padding(.bottom, 6)
            }
            ForEach(Array(hunks.enumerated()), id: \.offset) { _, hunk in
                let op = hunk.string("op", default: "ctx")
                Text((op == "add" ? "+ " : op == "del" ? "- " : "  ") + hunk.string("text"))
                    .font(.system(.footnote, design: .monospaced))
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 4)
                    .background(Self.bg(op))
                    .foregroundStyle(Self.fg(op))
            }
        }
        .padding(10)
        .background(Theme.field)
        .overlay(RoundedRectangle(cornerRadius: 10).stroke(Theme.stroke, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 10))
    }
    static func bg(_ op: String) -> Color {
        op == "add" ? Theme.green.opacity(0.14) : op == "del" ? Theme.coral.opacity(0.14) : .clear
    }
    static func fg(_ op: String) -> Color {
        op == "add" ? Theme.green : op == "del" ? Theme.coral : Theme.textSec
    }
}

File: cli-anr/app/src/main/java/com/aicodingremote/app/rendering/renderers/ContentRenderers.kt
package com.aicodingremote.app.rendering.renderers

import android.graphics.BitmapFactory
import android.util.Base64
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Photo
import androidx.compose.material3.Icon
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import com.aicodingremote.app.designsystem.Theme
import com.aicodingremote.app.models.Component
import com.aicodingremote.app.models.arrayValue
import com.aicodingremote.app.models.bool
import com.aicodingremote.app.models.double
import com.aicodingremote.app.models.doubleValue
import com.aicodingremote.app.models.get
import com.aicodingremote.app.models.string
import com.aicodingremote.app.models.stringValue

/** PROTOCOL §5.2 —— 普通文本 / markdown 文本。 */
@Composable
fun TextRenderer(component: Component) {
    val p = component.props
    val text = p.string("text")
    val isMarkdown = p.bool("markdown")
    if (isMarkdown) {
        MarkdownText(text)
    } else {
        val style = p.string("style", "body")
        val mono = p.bool("mono")
        val bold = p.bool("bold")
        val size = when (style) {
            "heading" -> 17.sp
            "caption" -> 13.sp
            else -> 15.sp
        }
        val weight = if (bold || style == "heading") FontWeight.SemiBold else FontWeight.Normal
        val fg: Color = p["color"]?.stringValue?.let { Theme.named(it, default = Theme.text) }
            ?: if (style == "caption") Theme.textSec else Theme.text
        val fill = p.bool("fill", default = true)
        Text(
            text = text,
            color = fg,
            fontSize = size,
            fontWeight = weight,
            fontFamily = if (mono) FontFamily.Monospace else null,
            modifier = if (fill) Modifier.fillMaxWidth() else Modifier,
        )
    }
}

/** PROTOCOL §5.2 —— 会话气泡。user → 右侧蓝气泡;agent → 左侧悬浮卡。 */
@Composable
fun BubbleRenderer(component: Component) {
    val p = component.props
    val text = p.string("text")
    val isMarkdown = p.bool("markdown")
    val isUser = p.string("role", "agent") == "user"

    if (isUser) {
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(16.dp))
                .background(Theme.blueBtn.copy(alpha = 0.22f))
                .padding(horizontal = 14.dp, vertical = 10.dp),
        ) {
            if (isMarkdown) MarkdownText(text)
            else Text(text, color = Theme.text, fontSize = 15.sp)
        }
    } else {
        val shape = RoundedCornerShape(Theme.rCard)
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(shape)
                .background(Theme.card, shape)
                .border(1.dp, Theme.stroke, shape)
                .padding(14.dp),
        ) {
            if (isMarkdown) MarkdownText(text)
            else Text(text, color = Theme.text, fontSize = 15.sp)
        }
    }
}

/** PROTOCOL §5.2 —— 代码块(标题栏 + 复制按钮 + 横向滚动单色代码)。 */
@Composable
fun CodeRenderer(component: Component) {
    val code = component.props.string("code")
    val language = component.props["language"]?.stringValue
    val copyable = component.props.bool("copyable", default = true)
    val shape = RoundedCornerShape(10.dp)

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Theme.field, shape)
            .border(1.dp, Theme.stroke, shape)
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        if (language != null || copyable) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                if (language != null) {
                    Text(language, color = Theme.textTer, fontSize = 11.sp)
                }
                Spacer(Modifier.weight(1f))
                if (copyable) {
                    CopyButton(text = code, label = "复制")
                }
            }
        }
        Row(Modifier.horizontalScroll(rememberScrollState())) {
            Text(
                text = code,
                color = Theme.text,
                fontFamily = FontFamily.Monospace,
                fontSize = 13.sp,
            )
        }
    }
}

/** PROTOCOL §5.2 —— 状态徽章(自由色名)。 */
@Composable
fun BadgeRenderer(component: Component) {
    val text = component.props.string("text")
    val color = Theme.named(component.props["color"]?.stringValue, default = Theme.textSec)
    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(color.copy(alpha = 0.16f))
            .padding(horizontal = 9.dp, vertical = 4.dp),
    ) {
        Text(text, color = color, fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
}

/** PROTOCOL §5.2 —— 键值对表格(左 secondary,右 primary)。 */
@Composable
fun KeyValueRenderer(component: Component) {
    val items = component.props["items"]?.arrayValue ?: emptyList()
    Column(
        Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        items.forEach { item ->
            Row(verticalAlignment = Alignment.Top, modifier = Modifier.fillMaxWidth()) {
                Text(item.string("k"), color = Theme.textSec, fontSize = 13.sp)
                Spacer(Modifier.weight(1f))
                Text(
                    item.string("v"),
                    color = Theme.text,
                    fontSize = 13.sp,
                    textAlign = TextAlign.End,
                )
            }
        }
    }
}

/**
 * PROTOCOL §5.2 —— 进度条。
 * - 有 `value`(0..1) → 线性进度条;
 * - 无 `value` → 圆形 spinner(对位 iOS 无限转圈)。
 */
@Composable
fun ProgressRenderer(component: Component) {
    val label = component.props["label"]?.stringValue
    val value = component.props.double("value") ?: component.props["value"]?.doubleValue
    Column(
        Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        if (label != null) Text(label, color = Theme.textSec, fontSize = 13.sp)
        if (value != null) {
            LinearProgressIndicator(
                progress = { value.toFloat().coerceIn(0f, 1f) },
                color = Theme.blue,
                trackColor = Theme.cardHi,
                modifier = Modifier.fillMaxWidth().height(6.dp),
            )
        } else {
            // 无限转圈(不知道何时结束):对位 iOS 默认 ProgressView() 的小转圈。
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                CircularProgressIndicator(
                    color = Theme.blue,
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(16.dp),
                )
            }
        }
    }
}

/**
 * PROTOCOL §5.2 —— 图片。三链路兼容(对位 iOS):
 * - `{id}`:控制通道只给 id,按 id 经 HTTP 拉(新链路,主流);
 * - `{data}`:内联 base64(旧链路);
 * - `{url}` / `{thumbUrl}`:外链(更旧)。
 */
@Composable
fun ImageRenderer(component: Component) {
    val p = component.props
    val shape = RoundedCornerShape(16.dp)
    val frame = Modifier
        .sizeIn(maxWidth = 240.dp, maxHeight = 300.dp)
        .shadow(6.dp, shape)
        .clip(shape)
        .border(0.5.dp, Theme.stroke.copy(alpha = 0.5f), shape)

    val id = p["id"]?.stringValue
    val dataStr = p["data"]?.stringValue
    val url = p["url"]?.stringValue ?: p["thumbUrl"]?.stringValue
    val alt = p["alt"]?.stringValue

    when {
        !id.isNullOrEmpty() -> {
            // 新链路:按 id 拉 — 复用 ChatRenderers 里的 IdAsyncImage(走相同的 LruCache)
            IdAsyncImage(id = id, maxW = 240.dp, maxH = 300.dp, contentDescription = alt)
        }
        dataStr != null -> {
            val bitmap = remember(dataStr) {
                runCatching {
                    val bytes = Base64.decode(dataStr, Base64.DEFAULT)
                    BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                }.getOrNull()
            }
            if (bitmap != null) {
                Image(
                    bitmap = bitmap.asImageBitmap(),
                    contentDescription = alt,
                    modifier = frame,
                    contentScale = ContentScale.Fit,
                    filterQuality = FilterQuality.Medium,
                )
            } else {
                ImagePlaceholder()
            }
        }
        url != null -> {
            AsyncImage(
                model = url,
                contentDescription = alt,
                modifier = frame,
                contentScale = ContentScale.Fit,
            )
        }
        else -> ImagePlaceholder()
    }
}

@Composable
private fun ImagePlaceholder() {
    Box(
        modifier = Modifier
            .size(width = 120.dp, height = 90.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(Theme.cardHi),
        contentAlignment = Alignment.Center,
    ) {
        Icon(Icons.Default.Photo, null, tint = Theme.textTer, modifier = Modifier.size(28.dp))
    }
}

/** PROTOCOL §5.2 —— 文件 diff(filename + hunks)。 */
@Composable
fun DiffRenderer(component: Component) {
    val filename = component.props["filename"]?.stringValue
    val hunks = component.props["hunks"]?.arrayValue ?: emptyList()
    val shape = RoundedCornerShape(10.dp)
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Theme.field, shape)
            .border(1.dp, Theme.stroke, shape)
            .padding(10.dp),
    ) {
        if (filename != null) {
            Text(
                filename,
                color = Theme.textSec,
                fontFamily = FontFamily.Monospace,
                fontSize = 12.sp,
                modifier = Modifier.padding(bottom = 6.dp),
            )
        }
        hunks.forEach { hunk ->
            val op = hunk.string("op", "ctx")
            val text = hunk.string("text")
            val (sign, fg, bg) = when (op) {
                "add" -> Triple("+ ", Theme.green, Theme.green.copy(alpha = 0.14f))
                "del" -> Triple("- ", Theme.coral, Theme.coral.copy(alpha = 0.14f))
                else -> Triple("  ", Theme.textSec, Color.Transparent)
            }
            Text(
                sign + text,
                color = fg,
                fontFamily = FontFamily.Monospace,
                fontSize = 12.sp,
                modifier = Modifier
                    .fillMaxWidth()
                    .background(bg)
                    .padding(horizontal = 4.dp),
            )
        }
    }
}
